src/hotspot/cpu/aarch64/universalUpcallHandler_aarch64.cpp

```diff
@@ -0,0 +1,187 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "asm/macroAssembler.hpp"
+#include "classfile/javaClasses.inline.hpp"
+#include "classfile/symbolTable.hpp"
+#include "include/jvm.h"
+#include "jni.h"
+#include "logging/log.hpp"
+#include "logging/logStream.hpp"
+#include "memory/allocation.inline.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/arrayOop.inline.hpp"
+#include "prims/universalUpcallHandler.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/javaCalls.hpp"
+#include "runtime/jniHandles.inline.hpp"
+
+static constexpr CodeBuffer::csize_t upcall_stub_size = 1024;
+
+extern struct JavaVM_ main_vm;
+
+static struct {
+  bool inited;
+  struct {
+    Klass* klass;
+    Symbol* name;
+    Symbol* sig;
+  } upcall_method;  // jdk.internal.foreign.abi.UniversalUpcallHandler::invoke
+} upcall_info;
+
+// FIXME: This should be initialized explicitly instead of lazily/racily
+static void upcall_init() {
+#if 0
+  fprintf(stderr, "upcall_init()\n");
+#endif
+
+  TRAPS = Thread::current();
+  ResourceMark rm;
+
+  const char* cname = "jdk/internal/foreign/abi/ProgrammableUpcallHandler";
+  const char* mname = "invoke";
+  const char* mdesc = "(Ljdk/internal/foreign/abi/ProgrammableUpcallHandler;J)V";
+  Symbol* cname_sym = SymbolTable::new_symbol(cname, (int)strlen(cname));
+  Symbol* mname_sym = SymbolTable::new_symbol(mname, (int)strlen(mname));
+  Symbol* mdesc_sym = SymbolTable::new_symbol(mdesc, (int)strlen(mdesc));
+
+#if 0
+  ::fprintf(stderr, "cname_sym: %p\n", cname_sym);
+  ::fprintf(stderr, "mname_sym: %p\n", mname_sym);
+  ::fprintf(stderr, "mdesc_sym: %p\n", mdesc_sym);
+#endif
+
+  Klass* k = SystemDictionary::resolve_or_null(cname_sym, THREAD);
+#if 0
+  ::fprintf(stderr, "Klass: %p\n", k);
+#endif
+
+  Method* method = k->lookup_method(mname_sym, mdesc_sym);
+#if 0
+  ::fprintf(stderr, "Method: %p\n", method);
+#endif
+
+  upcall_info.upcall_method.klass = k;
+  upcall_info.upcall_method.name = mname_sym;
+  upcall_info.upcall_method.sig = mdesc_sym;
+
+  upcall_info.inited = true;
+}
+
+static void upcall_helper(jobject rec, address buff) {
+  void *p_env = NULL;
+
+  Thread* thread = Thread::current_or_null();
+  if (thread == NULL) {
+    JavaVM_ *vm = (JavaVM *)(&main_vm);
+    vm -> functions -> AttachCurrentThreadAsDaemon(vm, &p_env, NULL);
+    thread = Thread::current();
+  }
+
+  assert(thread->is_Java_thread(), "really?");
+
+  ThreadInVMfromNative __tiv((JavaThread *)thread);
+
+  if (!upcall_info.inited) {
+    upcall_init();
+  }
+
+  ResourceMark rm;
+  JavaValue result(T_VOID);
+  JavaCallArguments args(2); // long = 2 slots
+
+  args.push_jobject(rec);
+  args.push_long((jlong) buff);
+
+  JavaCalls::call_static(&result, upcall_info.upcall_method.klass,
+                         upcall_info.upcall_method.name, upcall_info.upcall_method.sig,
+                         &args, thread);
+}
```

Comment:
This code shouldn't be in the cpu directory.  This should be in SharedRuntime or in jni.cpp.  It should have a JNI_ENTRY and not transition directly.  I don't know what AttachCurrentThreadAsDaemon does.

Reply:
Roger that.

We need the thread state transition though in case we get a random native thread calling us.

Reply:
yikes.  Does that work?